src/compiler/llvm/mir.rs

```diff
@@ -0,0 +1,215 @@
+//! Transforms the MIR representation into LLVM
+
+use std::collections::{hash_map::Entry, HashMap};
+
+use inkwell::{builder::Builder, context::Context, module::Module, values::*};
+
+use crate::{
+    compiler::{
+        mir::{ir::*, Transformer},
+        Span,
+    },
+    StringId, StringTable,
+};
+
+struct LlvmTransformer<'ctx> {
+    context: &'ctx Context,
+    module: Module<'ctx>,
+    builder: Builder<'ctx>,
+    table: &'ctx StringTable,
+    function: FunctionValue<'ctx>,
+
+    vars: HashMap<VarId, PointerValue<'ctx>>,
+}
+
+impl<'ctx> LlvmTransformer<'ctx> {
+    pub fn new(
+        func_name: StringId,
+        ctx: &'ctx Context,
+        module: &str,
+        table: &'ctx StringTable,
+    ) -> Self {
+        let module = ctx.create_module(module);
+
+        // Create a function to build
+        let ft = ctx.void_type().fn_type(&[], false);
+        let name = table.get(func_name).unwrap();
+        let function = module.add_function(&name, ft, None);
+
+        let builder = ctx.create_builder();
+
+        Self {
+            context: ctx,
+            module,
+            builder,
+            table,
+            function,
+            vars: HashMap::default(),
+        }
+    }
+
+    fn to_label(&self, vd: &VarDecl) -> String {
+        let name = self.table.get(vd.name()).unwrap();
+        let scope = vd.scope();
+        format!("{}_{}", name, scope)
+    }
+}
+
+impl<'ctx> Transformer<PointerValue<'ctx>, BasicValueEnum<'ctx>> for LlvmTransformer<'ctx> {
+    fn start_bb(&mut self, bb: BasicBlockId) {
+        let bb = self
+            .context
+            .append_basic_block(self.function, &bb.to_string());
+        self.builder.position_at_end(bb);
+    }
+
+    fn add_var(&mut self, id: VarId, decl: &VarDecl) {
+        let name = self.to_label(decl);
+
+        // Check if variable name already exists
+        match self.vars.entry(id) {
+            // If it does -> return an error
+            Entry::Occupied(_) => {
+                panic!("Attempting to add a variable that is already in the table")
+            }
```

Comment:
I think the XFormers should return Results for and the Traverser determines whether to fault or not.  The design ethos is that the Traverser is orchestrating the conversion and the Transformer is just trying to do what the Traverser asks.